openhcl/virt_mshv_vtl/src/processor/tdx/mod.rs

```diff
@@ -2394,6 +2369,93 @@ impl UhProcessor<'_, TdxBacked> {
+        // and the hardware did not generate an exit for a private/shared
+        // violation.
+        let address_type = self
+            .partition
+            .lower_vtl_memory_layout
+            .probe_address(canonical_gpa);
+
+        match address_type {
+            Some(AddressType::Mmio) => {
+                // Emulate the access.
+                self.emulate(
+                    dev,
+                    self.backing.vtls[intercepted_vtl]
+                        .interruption_information
+                        .valid(),
+                    intercepted_vtl,
+                    TdxEmulationCache::default(),
+                )
+                .await?;
+            }
+            Some(AddressType::Ram) => {
+                // TODO TDX: This path changes when we support VTL page
+                // protections and MNF. That will require injecting events to
+                // VTL1 or other handling.
+                //
+                // For now, we just check if the exit was suprious or if we
+                // should inject a machine check. An exit is considered spurious
+                // if the gpa is accessible.
+                if self.partition.gm[intercepted_vtl].check_gpa_readable(gpa) {
+                    tracelimit::warn_ratelimited!(gpa, "possible spurious EPT violation, ignoring");
+                } else {
+                    // TODO: It would be better to show what exact bitmap check
+                    // failed, but that requires some refactoring of how the
+                    // different bitmaps are stored. Do this when we support VTL
+                    // protections or MNF.
+                    //
+                    // If we entered this path, it means the bitmap check on
+                    // `check_gpa_readable` failed so we can assume that if the
+                    // address is shared, the actual state of the page is
+                    // private, and vice versa. This is because the address
+                    // should have already been checked to be valid memory
+                    // described to the guest or not.
+                    tracelimit::warn_ratelimited!(
+                        gpa,
+                        is_shared,
+                        ?ept_info,
+                        "guest accessed inaccessible gpa, injecting MC"
+                    );
+                    self.inject_mc(intercepted_vtl);
+                }
+            }
+            None => {
+                // The guest should never attempt to access address that are not
+                // described in mmio or ram. Inject a machine check.
+                tracelimit::warn_ratelimited!(
+                    gpa,
+                    is_shared,
+                    ?ept_info,
+                    "guest accessed gpa not described in memory layout, injecting MC"
+                );
+                self.inject_mc(intercepted_vtl);
```

Comment:
I worry this is too aggressive--it's not necessarily the case that all MMIO resides inside the MMIO ranges that are listed in the topology. E.g., I don't see how XAPIC works with this approach.

So I think either you need to weaken this check, or we need to do more work to ensure the topology is completely accurate.

Reply:
hmm... we don't support XAPIC today in TDX today, right, only x2apic? 

I think you are right though that this check is very strict. Note that this _is_ what our other paravisor implementation does. 

For non CVMs I am a bit worried about this being overzealous (and we probably don't need to be so strict). I bet we'd need to make this handling a bit more configurable - let me think about it some more.